src/binder/bind_node_visitor.cpp

```diff
@@ -197,6 +199,31 @@ void BindNodeVisitor::Visit(expression::FunctionExpression *expr) {
   // Visit the subtree first
   SqlNodeVisitor::Visit(expr);
 
+  // Specialize the first argument (DatePartType) for date functions, otherwise
+  // we have to do the string comparison to find out the corresponding
+  // DatePartType when scanning every tuple.
+  auto func_name = expr->GetFuncName();
+  if (func_name == "date_trunc" || func_name == "extract") {
```

Comment:
I didn't see the whole context here previously, but since you're working with `FunctionExpression`, you have access to the `OperatorId`. Can you use that instead of comparing strings? This way, we'll never have to worry about the case of strings, if it changes later.

Reply:
Yes. I think that would work!

Reply:
Actually I just tried and found out that it didn't work. The `OperatorId` was set after checking the name and input types of the function, which have to be set correctly for `Timestamp` functions here. (Basically we have to modify the `DatePartType` `Varchar` to `Integer` to pass the type checking to let the catalog find the `OperatorId` for the function.) So it seems to me has to be string comparison here.

Reply:
Can we put these in a hash table somewhere so that the will be easier to find instead of embedding them in the code like this? I suspect that there will be other functions that need to the same thing.

Reply:
Sounds reasonable. I can add a function in types.h called something like `FunctionNameToOperatorId()` to directly get the `OperatorId` through the name of a function.

Reply:
I feel like the mapping from function name to operator id should be in the catalog though.

Reply:
Actually while I was investigating how that should be done, I found there's already a mapping there for `BuiltInFunctions` used by the `Catalog`:
https://github.com/cmu-db/peloton/blob/master/src/catalog/catalog.cpp#L851
Assume we're only going to do this parameter specialization for built-in functions, this should be good. I can't really think of an example where we're going to manually infer the logic of a UDF and rewrite its parameters for now.